src/services/services.ts

```diff
@@ -4459,6 +4497,466 @@ namespace ts {
 
+        function getImportModuleCompletionsAtPosition(fileName: string, position: number): ImportCompletionInfo {
+            synchronizeHostData();
+
+            const sourceFile = getValidSourceFile(fileName);
+            if (isInReferenceComment(sourceFile, position)) {
+                return getTripleSlashReferenceCompletion(sourceFile, position);
+            }
+            else if (isInString(sourceFile, position)) {
+                const node = findPrecedingToken(position, sourceFile);
+                if (!node || node.kind !== SyntaxKind.StringLiteral) {
+                    return undefined;
+                }
+
+                if (node.parent.kind === SyntaxKind.ImportDeclaration || isExpressionOfExternalModuleImportEqualsDeclaration(node) || isRequireCall(node.parent, false)) {
+                    // Get all known external module names or complete a path to a module
+                    return {
+                        entries: getStringLiteralCompletionEntriesFromModuleNames(<StringLiteral>node),
+                        textSpan: getDirectoryFragmentTextSpan((<StringLiteral>node).text, node.getStart() + 1)
+                    };
+                }
+            }
+
+            return undefined;
+
+            function getStringLiteralCompletionEntriesFromModuleNames(node: StringLiteral) {
+                const literalValue = normalizeSlashes(node.text);
+
+                const scriptPath = node.getSourceFile().path;
+                const scriptDirectory = getDirectoryPath(scriptPath);
+                if (isPathRelativeToScript(literalValue) || isRootedDiskPath(literalValue)) {
+                    const compilerOptions = program.getCompilerOptions();
+                    if (compilerOptions.rootDirs) {
+                        return getCompletionEntriesForDirectoryFragmentWithRootDirs(
+                            compilerOptions.rootDirs, literalValue, scriptDirectory, getSupportedExtensions(program.getCompilerOptions()), /*includeExtensions*/false, scriptPath);
+                    }
+                    else {
+                        return getCompletionEntriesForDirectoryFragment(
+                            literalValue, scriptDirectory, getSupportedExtensions(program.getCompilerOptions()), /*includeExtensions*/false, scriptPath);
+                    }
+                }
+                else {
+                    // Check for node modules
+                    return getCompletionEntriesForNonRelativeModules(literalValue, scriptDirectory);
+                }
+            }
+
+            /**
+             * Takes a script path and returns paths for all potential folders that could be merged with its
+             * containing folder via the "rootDirs" compiler option
+             */
+            function getBaseDirectoriesFromRootDirs(rootDirs: string[], basePath: string, scriptPath: string, ignoreCase: boolean): string[] {
+                // Make all paths absolute/normalized if they are not already

```

Comment:
How expensive is this to do for each keypress in a project like Angular? Do we need to employ some type of caching mechanism rather than recalculate this each time?